project-koku/koku
# [COST-3083] - enable volume labels on OCP cost endpoint

## Jira Ticket

[COST-3083](https://issues.redhat.com/browse/COST-3083)

## Description

This change will add a new field to our OCP usage postgress table and populate with the combined volume + pod labels. We then also switch which field the provider_map is using making pod AND volume labels accessible from the OpenShift cost API.

## Testing

1. Checkout Branch
2. Let migration run
4. hit the following endpoint with volume and pod label group bys `/api/cost-management/v1/reports/openshift/costs/?group_by[tag:{volume_labels/pod_labels}]=*`

5. Hit the following endpoint:
```
http://localhost:8000/api/cost-management/v1/reports/openshift/costs/?group_by[tag:app]=*
```
6. Check the logs:
```
koku_server         | [2023-09-18 18:44:44,587] WARNING None 26 DEVELOPMENT is Enabled. Bypassing access lookup for user: {'identity': {'account_number': '10001', 'org_id': '1234567', 'type': 'User', 'user': {'username': 'user_dev', 'email': 'user_dev@foo.com', 'is_org_admin': 'True', 'access': {}}}, 'entitlements': {'cost_management': {'is_entitled': 'True'}}}
koku_server         | [2023-09-18 18:44:44,610] WARNING None 26 Returning default value for feature: cost-management.backend.override_customer_group_by_limit
koku_server         | [2023-09-18 18:44:44,610] WARNING None 26 Attempted to get feature_flag cost-management.backend.override_customer_group_by_limit, but client wasn't initialized!
koku_server         | [2023-09-18 18:44:44,638] WARNING None 26 Returning default value for feature: cost-management.backend.feature-cost-3083-all-labels
koku_server         | [2023-09-18 18:44:44,638] WARNING None 26 Attempted to get feature_flag cost-management.backend.feature-cost-3083-all-labels, but client wasn't initialized!
koku_server         | [2023-09-18 18:44:44,639] WARNING None 26 ('tag:app',) for costs has no entry in views. Using the default.
koku_server         | [2023-09-18 18:44:44,651] WARNING None 26 Exchange rates dictionary is not populated resulting in 'NoneType' object has no attribute 'currency_exchange_dictionary'.
```
Notice how substring `Returning default value for feature: cost-management.backend.feature-cost-3083-all-labels` is in the return.

7. Add a limit to the url:
```
http://localhost:8000/api/cost-management/v1/reports/openshift/costs/?group_by[tag:app]=*&limit=100
```
Check the logs:
```
koku_server         | [2023-09-18 18:45:31,652] WARNING None 26 DEVELOPMENT is Enabled. Bypassing access lookup for user: {'identity': {'account_number': '10001', 'org_id': '1234567', 'type': 'User', 'user': {'username': 'user_dev', 'email': 'user_dev@foo.com', 'is_org_admin': 'True', 'access': {}}}, 'entitlements': {'cost_management': {'is_entitled': 'True'}}}
koku_server         | [2023-09-18 18:45:31,752] WARNING None 26 Returning default value for feature: cost-management.backend.override_customer_group_by_limit
koku_server         | [2023-09-18 18:45:31,753] WARNING None 26 Attempted to get feature_flag cost-management.backend.override_customer_group_by_limit, but client wasn't initialized!
koku_server         | [2023-09-18 18:45:31,773] WARNING None 26 ('tag:app',) for costs has no entry in views. Using the default.
koku_server         | [2023-09-18 18:45:31,783] WARNING None 26 Exchange rates dictionary is not populated resulting in 'NoneType' object has no attribute 'currency_exchange_dictionary'.
```
Notice how `Returning default value for feature: cost-management.backend.feature-cost-3083-all-labels` does not appear in the logs now.

8. Log into unleash:
```
http://localhost:4242/
```
9. Add `org1234567` the `cost-management.backend.feature-cost-3083-all-labels` flag.


## Notes

...
